radix-engine/src/wasm/wasmi.rs

```diff
@@ -11,279 +14,478 @@ use crate::wasm::constants::*;
 use crate::wasm::errors::*;
 use crate::wasm::traits::*;
 
+type HostState = WasmiInstanceEnv;
+
+/// A `WasmiModule` defines a parsed WASM module "template" Instance (with imports already defined) and Store,
+/// which keeps user data.
+/// "Template" (Store, Instance) tuple are cached together, and never to be invoked.
+/// Upon instantiation Instance and Store are cloned, so the state is not shared between instances.
+/// It is correctly `Send + Sync` - which is good, because this is the thing which is cached in the
+/// ScryptoInterpreter caches.
```

Comment:
It's correctly `Send + Sync` under the assumption that the `data` in the `Store` is set to a valid value unpon invocation.

It should not be sent from one thread to another, because the pointer may point to volatile data.

Reply:
Why does `WasmiModule` contain an `Instance` or `State`? Doesn't this mean that it shouldn't be cached?

Can we explain why it is safe to cache the `WasmiModule` even if it contains `store` and `instance`?